activerecord/test/cases/adapters/mysql2/mysql2_adapter_test.rb

```diff
@@ -18,6 +18,12 @@ def test_connection_error
     assert_kind_of ActiveRecord::ConnectionAdapters::NullPool, error.connection_pool
   end
 
+  def test_connection_error_via_socket_on_localhost
+    assert_raises ActiveRecord::CannotConnectThroughSocketOnLocalhost do
+      ActiveRecord::ConnectionAdapters::Mysql2Adapter.new(host: "localhost", prepared_statements: false).connect!
```

Comment:
Does this test fails when you have mysql running on your local machine outside a container?

Reply:
No, they work as intended, @rafaelfranca:

```shell
ruy@devastator ~/G/r/activerecord (feature/new-exception-when-connecting-to-mysql-running-on-localhost)> bundle exec rake test TEST="test/cases/adapters/mysql2/mysql2_adapter_test.rb"
/home/ruy/.asdf/installs/ruby/3.2.3/bin/ruby -w -I"lib:test" /home/ruy/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/rake-13.1.0/lib/rake/rake_test_loader.rb "test/cases/adapters/mysql2/mysql2_adapter_test.rb"
Using mysql2
Run options: --seed 6656

# Running:

..............S.......

Finished in 1.328510s, 16.5599 runs/s, 40.6470 assertions/s.
22 runs, 54 assertions, 0 failures, 0 errors, 1 skips

You have skipped tests. Run with --verbose for details.
/home/ruy/.asdf/installs/ruby/3.2.3/bin/ruby -w -I"lib:test" /home/ruy/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/rake-13.1.0/lib/rake/rake_test_loader.rb "test/cases/adapters/mysql2/mysql2_adapter_test.rb"
Using mysql2
Run options: --seed 56615

# Running:

...................S..

Finished in 1.300546s, 16.9160 runs/s, 41.5210 assertions/s.
22 runs, 54 assertions, 0 failures, 0 errors, 1 skips
```

Here is what's going on behind the scenes:

```shell
ruy@devastator ~> service mariadb status
Redirecting to /bin/systemctl status mariadb.service
● mariadb.service - MariaDB 10.5 database server
     Loaded: loaded (/usr/lib/systemd/system/mariadb.service; disabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: active (running) since Mon 2024-01-29 22:08:55 -03; 14min ago
       Docs: man:mariadbd(8)
             https://mariadb.com/kb/en/library/systemd/
    Process: 169568 ExecStartPre=/usr/libexec/mariadb-check-socket (code=exited, status=0/SUCCESS)
    Process: 169593 ExecStartPre=/usr/libexec/mariadb-prepare-db-dir mariadb.service (code=exited, status=0/SUCCESS)
    Process: 169735 ExecStartPost=/usr/libexec/mariadb-check-upgrade (code=exited, status=0/SUCCESS)
   Main PID: 169703 (mariadbd)
     Status: "Taking your SQL requests now..."
      Tasks: 13 (limit: 18259)
     Memory: 872.6M
        CPU: 8.466s
     CGroup: /system.slice/mariadb.service
             └─169703 /usr/libexec/mariadbd --basedir=/usr

Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: you need to be the system 'mysql' user to connect.
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: After connecting you can set the password, if you would need to be
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: able to connect as any of these users with a password and without sudo
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: See the MariaDB Knowledgebase at https://mariadb.com/kb
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: Please report any problems at https://mariadb.org/jira
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: The latest information about MariaDB is available at https://mariadb.org/.
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: Consider joining MariaDB's strong and vibrant community:
Jan 29 22:08:55 devastator mariadb-prepare-db-dir[169643]: https://mariadb.org/get-involved/
Jan 29 22:08:55 devastator (mariadbd)[169703]: mariadb.service: Referenced but unset environment variable evaluates to an empty string: MYSQLD_OPTS, _WSREP_NEW_CLUSTER
Jan 29 22:08:55 devastator systemd[1]: Started mariadb.service - MariaDB 10.5 database server.
ruy@devastator ~> strace -o mysql_connection.txt mysql --user rails --host localhost
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 181
Server version: 10.5.23-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> ^DBye
ruy@devastator ~> rg 'mysql.sock' mysql_connection.txt
279:connect(3, {sa_family=AF_UNIX, sun_path="/var/lib/mysql/mysql.sock"}, 110) = 0
ruy@devastator ~> strace -o mysql_connection.txt mysql --user rails --host 127.0.0.1
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 182
Server version: 10.5.23-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> ^DBye
ruy@devastator ~> rg 'mysql.sock' mysql_connection.txt
ruy@devastator ~ [1]> rg '127.0.0.1' mysql_connection.txt
1:execve("/usr/bin/mysql", ["mysql", "--user", "rails", "--host", "127.0.0.1"], 0xffffc6f8c210 /* 50 vars */) = 0
278:connect(3, {sa_family=AF_INET, sin_port=htons(3306), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EINPROGRESS (Operation now in progress)
ruy@devastator ~>
```